### PR TITLE
new MMTF format (#907)

### LINKED ISSUE #907: support mmtf format
@arose [suggested on twitter](https://twitter.com/asrmoin/status/753430985765588992) to use [MMTF](http://mmtf.rcsb.org/)  for storing big structures. "The Macromolecular Transmission Format (MMTF) is a new compact binary format to transmit and store biomolecular structural data quickly and accurately."

There is a python implementation available: https://github.com/rcsb/mmtf-python

### PR DESCRIPTION
**Pull request by richardjgowers**
Fixes #907

Changes made in this Pull Request:
- Adds support for MMTF format
## PR Checklist
- [ ] Tests?
- [ ] Docs?
- [ ] CHANGELOG updated?
- [ ] Issue raised/referenced?

Works with 4CUP file from mmtf tests (found here https://github.com/rcsb/mmtf-python/tree/master/mmtf/tests/testdatastore)

There's a few things still left to do, but I'd say this is nearly done.  I wasn't sure how segments (chains?) worked, but admittedly didn't spend long on this.

Requires the mmtf package (https://github.com/rcsb/mmtf-python) I'm not sure how requirements for this are.  [This says](https://github.com/rcsb/mmtf-python/blob/master/requirements.txt) it requires ipython, but searching the repo I can't find any reference to it?

For the tests, I wasn't sure if there was a biothing? that we could use which we already have PDB tests for?  The structures should be identical in theory?

### CODE REVIEW

**Comment by orbeckst** [comment]
The AdK structure in the PSF/DCD tests is based on [1AKE](http://www.rcsb.org/pdb/explore.do?structureId=1ake), chain A. 
```
import mmtf
decoded_data = mmtf.fetch('1AKE')
```

I didn't spend enough time with mmtf to figure out how to access chain A, though.... but:The [mmtf format descriptiob v0.2](https://github.com/rcsb/mmtf/blob/v0.2/spec.md) is enlightening. It seems that one has to [explicitly build the offsets in the coordinate arrays from chain data](https://github.com/rcsb/mmtf/blob/v0.2/spec.md#chain-data) as shown for [groups_per_chain](https://github.com/rcsb/mmtf/blob/v0.2/spec.md#groupsperchain).

Getting at residues appears to require [group_type_list](https://github.com/rcsb/mmtf/blob/v0.2/spec.md#grouptypelist), which contains indices into  [group_list](https://github.com/rcsb/mmtf/blob/v0.2/spec.md#grouplist)... @richardjgowers and @dotsdl should feel at home here, looks like #363... Basically, `group_type_list` is something like `ag.residues`. `group_id_list` is `ag.residues.resnums`. To get all residue data for resid `n`, do `group_list[group_type_list[n]]` and to get the resnum, do `group_id_list[n]`.

Oh well, and I haven't even looked at `altLoc` yet... have fun :-).

But wait: in [Traversal](https://github.com/rcsb/mmtf/blob/v0.2/spec.md#traversal) they show how to unpack everything... maybe I should have read first and type later? ;-)

**Comment by orbeckst** [comment]
Don't forget to add mmtf as a dependency in `setup.py`.

File: package/MDAnalysis/coordinates/MMTF.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+        # TOOD: Check units?
+        top = mmtf.parse(self.filename)
+        self.n_atoms = top.num_atoms
+
+        self.ts = ts = self._Timestep(self.n_atoms,
+                                      **self._ts_kwargs)
+        ts._pos[:, 0] = top.x_coord_list.copy()
+        ts._pos[:, 1] = top.y_coord_list.copy()
+        ts._pos[:, 2] = top.z_coord_list.copy()
+        ts.dimensions = top.unit_cell
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/MMTF.py]: To be consistent, this should be a copy, too. At least that will allow `top` to be garbage collected.

File: package/MDAnalysis/coordinates/MMTF.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    format = 'MMTF'
+
+    def _read_first_frame(self):
+        # TOOD: Check units?
+        top = mmtf.parse(self.filename)
+        self.n_atoms = top.num_atoms
+
+        self.ts = ts = self._Timestep(self.n_atoms,
+                                      **self._ts_kwargs)
+        ts._pos[:, 0] = top.x_coord_list.copy()
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/MMTF.py]: Do you need a copy here? Isn't the slice assignment already making a copy?

File: package/MDAnalysis/coordinates/MMTF.py (hunk 1 above)
Comment (orbeckst) [inline package/MDAnalysis/coordinates/MMTF.py]: Given that `ts.dimensions` is one of the attributes that MDAnalysis guarantees a format for, I would unpack `top.unit_cell` explicitly. Or at least add a comment that the MMTF format is the same as the MDA one.

File: package/MDAnalysis/coordinates/MMTF.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+        ts._pos[:, 1] = top.y_coord_list.copy()
+        ts._pos[:, 2] = top.z_coord_list.copy()
+        ts.dimensions = top.unit_cell
+
+        ts.data['occupancy'] = top.occupancy_list.copy()
```
Comment (orbeckst) [inline package/MDAnalysis/coordinates/MMTF.py]: Is this going to be a float64 array? What does the PDBReader do? Should be the same type.

I'd be careful with wholesale copying of "foreign" data structures.

**Comment by richardjgowers** [comment]
Thanks @orbeckst

I had a question about these files (and PDB/bio files in general), they can hold multiple "models" which PDB uses to represent frames in a trajectory.  Will all models always have the same number of atoms, or can different models have different atoms?

**Comment by orbeckst** [comment]
> On 27 Jul, 2016, at 13:33, Richard Gowers <notifications@github.com> wrote:
> 
> I had a question about these files (and PDB/bio files in general), they can hold multiple "models" which PDB uses to represent frames in a trajectory.

Actually, the protein databank does not care about trajectories. Some experimental techniques such as NMR produce multiple models. We just co-opt the format.

> Will all models always have the same number of atoms, or can different models have different atoms?

I am pretty sure that they are supposed to have the same number of atoms  but the PDB specs on MODEL records may say more http://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#MODEL :

"All models in an NMR ensemble should be homogeneous – each model should have the exact same atoms (hydrogen and heavy atoms), sequence and chemistry."

The answer to your question is YES.


--
Oliver Beckstein * orbeckst@gmx.net
skype: orbeckst  * orbeckst@gmail.com

**Pull request closed without merging**